File: server/app/models/transaction.py
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, Optional

from sqlalchemy import DateTime
from sqlmodel import Field, Relationship

from app.models.base import BaseModel, SoftDeleteMixin

if TYPE_CHECKING:
    from app.models.user import User


class CategoryEnum(str, Enum):
    """Predefined spending categories for transaction classification."""

    INCOME = "Income"
    TRANSPORTATION = "Transportation"
    SALARY = "Salary"
    HOUSEHOLD_UTILITIES = "Household & Utilities"
    TAX_FINES = "Tax & Fines"
    MISCELLANEOUS = "Miscellaneous"
    FOOD_GROCERIES = "Food & Groceries"
    FOOD_DELIVERY = "Food Delivery"
    ATM = "ATM"
    INSURANCE = "Insurances"
    SHOPPING = "Shopping"
    BARS_RESTAURANTS = "Bars & Restaurants"
    EDUCATION = "Education"
    FAMILY_FRIENDS = "Family & Friends"
    DONATIONS_CHARITY = "Donations & Charity"
    HEALTHCARE_DRUG_STORES = "Healthcare & Drug Stores"
    LEISURE_ENTERTAINMENT = "Leisure & Entertainment"
    MEDIA_ELECTRONICS = "Media & Electronics"
    SAVINGS_INVESTMENTS = "Savings & Investments"
    TRAVEL_HOLIDAYS = "Travel & Holidays"


class Transaction(BaseModel, SoftDeleteMixin, table=True):
    """Financial transaction record owned by a user.

    confidence_score=1.0 signals a manually-entered (human-verified) transaction.
    Soft-deleted records (deleted_at IS NOT NULL) are excluded from all queries.
    """

    # Primary & Foreign Keys
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: str = Field(foreign_key="user.id", index=True)

    # Core Transaction Data
    date: datetime = Field(sa_type=DateTime(timezone=True), index=True)
    merchant: str
    amount: float
    description: Optional[str] = None
    original_category: Optional[str] = None

    # Categorization
    category: CategoryEnum = Field(index=True)
    confidence_score: float = Field(ge=0.0, le=1.0)

    # Recurring Detection
    is_recurring: bool = Field(default=False)

    # Relationships
    user: Optional["User"] = Relationship(back_populates="transactions")

File: web-app/src/types/transaction.ts
export type CategoryEnum =
  | 'Income'
  | 'Transportation'
  | 'Salary'
  | 'Household & Utilities'
  | 'Tax & Fines'
  | 'Miscellaneous'
  | 'Food & Groceries'
  | 'Food Delivery'
  | 'ATM'
  | 'Insurances'
  | 'Shopping'
  | 'Bars & Restaurants'
  | 'Education'
  | 'Family & Friends'
  | 'Donations & Charity'
  | 'Healthcare & Drug Stores'
  | 'Leisure & Entertainment'
  | 'Media & Electronics'
  | 'Savings & Investments'
  | 'Travel & Holidays'

export const CATEGORY_OPTIONS: CategoryEnum[] = [
  'Income',
  'Transportation',
  'Salary',
  'Household & Utilities',
  'Tax & Fines',
  'Miscellaneous',
  'Food & Groceries',
  'Food Delivery',
  'ATM',
  'Insurances',
  'Shopping',
  'Bars & Restaurants',
  'Education',
  'Family & Friends',
  'Donations & Charity',
  'Healthcare & Drug Stores',
  'Leisure & Entertainment',
  'Media & Electronics',
  'Savings & Investments',
  'Travel & Holidays',
]

export interface TransactionResponse {
  id: number
  user_id: string
  /** ISO 8601 date string — do NOT use Date type */
  date: string
  merchant: string
  /** CRITICAL: Decimal serializes as string in JSON — must NOT be number */
  amount: string
  description: string | null
  original_category: string | null
  category: CategoryEnum
  confidence_score: number
  is_recurring: boolean
  /** ISO 8601 datetime string */
  created_at: string
  /** ISO 8601 datetime string */
  updated_at: string
}

export interface TransactionListResponse {
  items: TransactionResponse[]
  total: number
  offset: number
  limit: number
}

export interface TransactionFilters {
  date_from?: string
  date_to?: string
  category?: CategoryEnum
  merchant?: string
  amount_min?: string
  amount_max?: string
  is_recurring?: boolean
  sort_by?: 'date' | 'amount' | 'merchant'
  sort_order?: 'asc' | 'desc'
  offset?: number
  limit?: number
}

export interface BatchUpdateItem {
  id: number
  category?: CategoryEnum
}

/** Maximum 100 items per request */
export interface BatchUpdateRequest {
  items: BatchUpdateItem[]
}

export interface BatchUpdateResponse {
  updated: number
}

/** Maximum 100 ids per request */
export interface BatchDeleteRequest {
  ids: number[]
}

export interface BatchDeleteResponse {
  deleted: number
}

export interface CreateTransactionRequest {
  /** ISO 8601 date string, e.g. '2024-01-15' */
  date: string
  merchant: string
  /** Decimal as string — e.g. '-42.50' for expense, '1200.00' for income */
  amount: string
  category: CategoryEnum
  description?: string
  is_recurring?: boolean
}

export interface TransactionUpdateRequest {
  date?: string
  merchant?: string
  /** CRITICAL: Decimal as string — e.g. '-42.50' for expense, '1200.00' for income; must NOT be number */
  amount?: string
  /** null not supported for clearing — backend uses exclude_none=True */
  description?: string
  category?: CategoryEnum
  is_recurring?: boolean
}
